File: zipkin-server/src/test/java/zipkin2/server/internal/ui/ZipkinUiConfigurationTest.java
/*
 * Copyright The OpenZipkin Authors
 * SPDX-License-Identifier: Apache-2.0
 */
package zipkin2.server.internal.ui;

import com.linecorp.armeria.common.AggregatedHttpResponse;
import com.linecorp.armeria.common.HttpHeaderNames;
import com.linecorp.armeria.common.HttpMethod;
import com.linecorp.armeria.common.HttpRequest;
import com.linecorp.armeria.common.MediaType;
import com.linecorp.armeria.common.RequestHeaders;
import com.linecorp.armeria.server.HttpService;
import com.linecorp.armeria.server.ServiceRequestContext;
import io.netty.handler.codec.http.cookie.ClientCookieEncoder;
import io.netty.handler.codec.http.cookie.Cookie;
import io.netty.handler.codec.http.cookie.DefaultCookie;
import java.io.ByteArrayInputStream;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.BeanCreationException;
import org.springframework.beans.factory.NoSuchBeanDefinitionException;
import org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration;
import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;
import org.springframework.core.io.ClassPathResource;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatExceptionOfType;

class ZipkinUiConfigurationTest {

  AnnotationConfigApplicationContext context;

  @AfterEach void close() {
    if (context != null) {
      context.close();
    }
  }

  @Test void indexContentType() {
    context = createContext();
    assertThat(
      serveIndex().headers().contentType())
      .isEqualTo(MediaType.HTML_UTF_8);
  }

  @Test void indexHtml() {
    // Instantiate directly so that spring doesn't cache it
    ZipkinUiConfiguration ui = new ZipkinUiConfiguration();
    ui.ui = new ZipkinUiProperties();
    ui.lensIndexHtml = new ClassPathResource("does-not-exist.html");
    assertThatExceptionOfType(BeanCreationException.class).isThrownBy(ui::indexService);
  }

  @Test void canOverridesProperty_defaultLookback() {
    context = createContextWithOverridenProperty("zipkin.ui.defaultLookback:100");

    assertThat(context.getBean(ZipkinUiProperties.class).getDefaultLookback())
      .isEqualTo(100);
  }

  @Test void canOverrideProperty_logsUrl() {
    final String url = "http://mycompany.com/kibana";
    context = createContextWithOverridenProperty("zipkin.ui.logs-url:" + url);

    assertThat(context.getBean(ZipkinUiProperties.class).getLogsUrl()).isEqualTo(url);
  }

  @Test void canOverrideProperty_archivePostUrl() {
    final String url = "http://zipkin.archive.com/api/v2/spans";
    context = createContextWithOverridenProperty("zipkin.ui.archive-post-url:" + url);

    assertThat(context.getBean(ZipkinUiProperties.class).getArchivePostUrl()).isEqualTo(url);
  }

  @Test void canOverrideProperty_archiveUrl() {
    final String url = "http://zipkin.archive.com/zipkin/traces/{traceId}";
    context = createContextWithOverridenProperty("zipkin.ui.archive-url:" + url);

    assertThat(context.getBean(ZipkinUiProperties.class).getArchiveUrl()).isEqualTo(url);
  }

  @Test void canOverrideProperty_supportUrl() {
    final String url = "http://mycompany.com/file-a-bug";
    context = createContextWithOverridenProperty("zipkin.ui.support-url:" + url);

    assertThat(context.getBean(ZipkinUiProperties.class).getSupportUrl()).isEqualTo(url);
  }

  @Test void logsUrlIsNullIfOverridenByEmpty() {
    context = createContextWithOverridenProperty("zipkin.ui.logs-url:");

    assertThat(context.getBean(ZipkinUiProperties.class).getLogsUrl()).isNull();
  }

  @Test void logsUrlIsNullByDefault() {
    context = createContext();

    assertThat(context.getBean(ZipkinUiProperties.class).getLogsUrl()).isNull();
  }

  @Test void canOverridesProperty_disable() {
    assertThatExceptionOfType(NoSuchBeanDefinitionException.class).isThrownBy(() -> {
      context = createContextWithOverridenProperty("zipkin.ui.enabled:false");

      context.getBean(ZipkinUiProperties.class);
    });
  }

  @Test void canOverridesProperty_searchEnabled() {
    context = createContextWithOverridenProperty("zipkin.ui.search-enabled:false");

    assertThat(context.getBean(ZipkinUiProperties.class).isSearchEnabled()).isFalse();
  }

  @Test void canOverridesProperty_dependenciesEnabled() {
    context = createContextWithOverridenProperty("zipkin.ui.dependency.enabled:false");

    assertThat(context.getBean(ZipkinUiProperties.class).getDependency().isEnabled()).isFalse();
  }

  @Test void canOverrideProperty_dependencyLowErrorRate() {
    context = createContextWithOverridenProperty("zipkin.ui.dependency.low-error-rate:0.1");

    assertThat(context.getBean(ZipkinUiProperties.class).getDependency().getLowErrorRate())
      .isEqualTo(0.1f);
  }

  @Test void canOverrideProperty_dependencyHighErrorRate() {
    context = createContextWithOverridenProperty("zipkin.ui.dependency.high-error-rate:0.1");

    assertThat(context.getBean(ZipkinUiProperties.class).getDependency().getHighErrorRate())
      .isEqualTo(0.1f);
  }

  @Test void defaultBaseUrl_doesNotChangeResource() {
    context = createContext();

    assertThat(new ByteArrayInputStream(serveIndex().content().array()))
      .hasSameContentAs(getClass().getResourceAsStream("/zipkin-lens/index.html"));
  }

  @Test void canOverrideProperty_basePath() {
    context = createContextWithOverridenProperty("zipkin.ui.basepath:/admin/zipkin");

    assertThat(serveIndex().contentUtf8()).isEqualTo("""
      <!-- simplified version of /zipkin-lens/index.html -->
      <html>
        <head>
          <base href="/admin/zipkin/">
          <link rel="icon" href="./favicon.ico">
          <script type="module" crossorigin="" src="./static/js/index.js"></script>
          <link rel="stylesheet" href="./static/css/index.css">
        </head>
        <body>zipkin-lens</body>
      </html>
      """
    );
  }

  @Test void lensCookieOverridesIndex() {
    context = createContext();

    assertThat(serveIndex(new DefaultCookie("lens", "true")).contentUtf8())
      .contains("zipkin-lens");
  }

  @Test void canOverrideProperty_root() {
    context = createContextWithOverridenProperty("zipkin.ui.basepath:/");

    assertThat(serveIndex().contentUtf8()).isEqualTo("""
      <!-- simplified version of /zipkin-lens/index.html -->
      <html>
        <head>
          <base href="/">
          <link rel="icon" href="./favicon.ico">
          <script type="module" crossorigin="" src="./static/js/index.js"></script>
          <link rel="stylesheet" href="./static/css/index.css">
        </head>
        <body>zipkin-lens</body>
      </html>
      """
    );
  }

  AggregatedHttpResponse serveIndex(Cookie... cookies) {
    RequestHeaders headers = RequestHeaders.of(HttpMethod.GET, "/");
    String encodedCookies = ClientCookieEncoder.LAX.encode(cookies);
    if (encodedCookies != null) {
      headers = headers.toBuilder().set(HttpHeaderNames.COOKIE, encodedCookies).build();
    }
    HttpRequest req = HttpRequest.of(headers);
    try {
      return context.getBean(HttpService.class)
        .serve(ServiceRequestContext.of(req), req).aggregate()
        .get();
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  private static AnnotationConfigApplicationContext createContext() {
    AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext();
    context.register(PropertyPlaceholderAutoConfiguration.class, ZipkinUiConfiguration.class);
    context.refresh();
    return context;
  }

  private static AnnotationConfigApplicationContext createContextWithOverridenProperty(
    String pair) {
    AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext();
    TestPropertyValues.of(pair).applyTo(context);
    context.register(PropertyPlaceholderAutoConfiguration.class, ZipkinUiConfiguration.class);
    context.refresh();
    return context;
  }
}

File: zipkin-server/src/test/kotlin/zipkin2/server/internal/ui/ZipkinUiConfigurationTest.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package zipkin2.server.internal.ui

import com.linecorp.armeria.common.AggregatedHttpMessage
import com.linecorp.armeria.common.HttpHeaderNames
import com.linecorp.armeria.common.HttpMethod
import com.linecorp.armeria.common.HttpRequest
import com.linecorp.armeria.common.MediaType
import com.linecorp.armeria.common.RequestHeaders
import com.linecorp.armeria.server.ServiceRequestContext
import io.netty.handler.codec.http.cookie.ClientCookieEncoder
import io.netty.handler.codec.http.cookie.Cookie
import io.netty.handler.codec.http.cookie.DefaultCookie
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Test
import org.springframework.beans.factory.NoSuchBeanDefinitionException
import org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration
import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.annotation.AnnotationConfigApplicationContext
import org.springframework.core.io.ClassPathResource
import java.io.ByteArrayInputStream
import java.io.FileNotFoundException

class ZipkinUiConfigurationTest {
  val context = AnnotationConfigApplicationContext()
  @After fun closeContext() = context.close()

  @Test fun indexHtmlFromClasspath() {
    refreshContext();

    assertThat(context.getBean(ZipkinUiConfiguration::class.java).indexHtml)
      .isNotNull
  }

  @Test fun indexContentType() {
    refreshContext()

    assertThat(serveIndex().headers().contentType())
      .isEqualTo(MediaType.HTML_UTF_8)
  }

  @Test fun invalidIndexHtml() {
    TestPropertyValues.of("zipkin.ui.basepath:/foo/bar").applyTo(context)
    refreshContext()

    // I failed to make Jsoup barf, even on nonsense like: "<head wait no I changed my mind this HTML is totally invalid <<<<<<<<<<<"
    // So let's just run with a case where the file doesn't exist
    val ui = context.getBean(ZipkinUiConfiguration::class.java)
    ui.indexHtml = ClassPathResource("does-not-exist.html")

    try {
      serveIndex()

      assertThat(false).isTrue()
    } catch (e: RuntimeException) {
      assertThat(e).hasRootCauseInstanceOf(FileNotFoundException::class.java)
    }
  }

  @Test fun canOverridesProperty_defaultLookback() {
    TestPropertyValues.of("zipkin.ui.defaultLookback:100").applyTo(context)
    refreshContext()

    assertThat(context.getBean(ZipkinUiProperties::class.java).defaultLookback)
      .isEqualTo(100)
  }

  @Test fun canOverrideProperty_logsUrl() {
    val url = "http://mycompany.com/kibana"
    TestPropertyValues.of("zipkin.ui.logs-url:$url").applyTo(context)
    refreshContext()

    assertThat(context.getBean(ZipkinUiProperties::class.java).logsUrl).isEqualTo(url)
  }

  @Test fun logsUrlIsNullIfOverridenByEmpty() {
    TestPropertyValues.of("zipkin.ui.logs-url:").applyTo(context)
    refreshContext()

    assertThat(context.getBean(ZipkinUiProperties::class.java).logsUrl).isNull()
  }

  @Test fun logsUrlIsNullByDefault() {
    refreshContext()

    assertThat(context.getBean(ZipkinUiProperties::class.java).logsUrl).isNull()
  }

  @Test(expected = NoSuchBeanDefinitionException::class)
  fun canOverridesProperty_disable() {
    TestPropertyValues.of("zipkin.ui.enabled:false").applyTo(context)
    refreshContext()

    context.getBean(ZipkinUiProperties::class.java)
  }

  @Test fun canOverridesProperty_searchEnabled() {
    TestPropertyValues.of("zipkin.ui.search-enabled:false").applyTo(context)
    refreshContext()

    assertThat(context.getBean(ZipkinUiProperties::class.java).isSearchEnabled).isFalse()
  }

  @Test fun canOverrideProperty_dependencyLowErrorRate() {
    TestPropertyValues.of("zipkin.ui.dependency.low-error-rate:0.1").applyTo(context)
    refreshContext()

    assertThat(context.getBean(ZipkinUiProperties::class.java).dependency.lowErrorRate)
      .isEqualTo(0.1f)
  }

  @Test fun canOverrideProperty_dependencyHighErrorRate() {
    TestPropertyValues.of("zipkin.ui.dependency.high-error-rate:0.1").applyTo(context)
    refreshContext()

    assertThat(context.getBean(ZipkinUiProperties::class.java).dependency.highErrorRate)
      .isEqualTo(0.1f)
  }

  @Test fun defaultBaseUrl_doesNotChangeResource() {
    refreshContext()

    assertThat(ByteArrayInputStream(serveIndex().content().array()))
      .hasSameContentAs(javaClass.getResourceAsStream("/zipkin-ui/index.html"))
  }

  @Test fun canOverideProperty_basePath() {
    TestPropertyValues.of("zipkin.ui.basepath:/foo/bar").applyTo(context)
    refreshContext()

    assertThat(serveIndex().contentUtf8())
      .contains("<base href=\"/foo/bar/\">")
  }

  @Test fun lensCookieOverridesIndex() {
    refreshContext()

    assertThat(serveIndex(DefaultCookie("lens", "true")).contentUtf8())
      .contains("zipkin-lens")
  }

  @Test fun canOverideProperty_specialCaseRoot() {
    TestPropertyValues.of("zipkin.ui.basepath:/").applyTo(context)
    refreshContext()

    assertThat(serveIndex().contentUtf8())
      .contains("<base href=\"/\">")
  }

  private fun serveIndex(vararg cookies: Cookie): AggregatedHttpMessage {
    val headers = RequestHeaders.builder(HttpMethod.GET, "/")
    val encodedCookies = ClientCookieEncoder.LAX.encode(*cookies)
    if (encodedCookies != null) {
      headers.set(HttpHeaderNames.COOKIE, encodedCookies)
    }
    val req = HttpRequest.of(headers.build())
    return context.getBean(ZipkinUiConfiguration::class.java).indexSwitchingService()
      .serve(ServiceRequestContext.of(req), req).aggregate()
      .get()
  }

  private fun refreshContext() {
    context.register(
      PropertyPlaceholderAutoConfiguration::class.java,
      ZipkinUiConfiguration::class.java)
    context.refresh()
  }
}
